tests/unit/moduleapi/moduleconfigs.tcl

```diff
@@ -0,0 +1,165 @@
+        catch {[r config set moduleconfigs.numeric -10]} e
+        assert_match {*argument must be between*} $e
+    }
+
+    test {Enums only able to be set to passed in values} {
+        # Module authors specify what values are valid for enums, check that only those values are ok on a set
+        catch {[r config set moduleconfigs.enum four]} e
+        assert_match {*argument must be one of the following*} $e
+    }
+
+    test {Unload removes module configs} {
+        r module unload moduleconfigs
+        assert_equal [r config get moduleconfigs.*] ""
+        r module load $testmodule
+        # these should have reverted back to their module specified values
+        assert_equal [r config get moduleconfigs.mutable_bool] "moduleconfigs.mutable_bool yes"
+        assert_equal [r config get moduleconfigs.immutable_bool] "moduleconfigs.immutable_bool no"
+        assert_equal [r config get moduleconfigs.memory_numeric] "moduleconfigs.memory_numeric 1024"
+        assert_equal [r config get moduleconfigs.string] "moduleconfigs.string log4j"
+        assert_equal [r config get moduleconfigs.enum] "moduleconfigs.enum one"
+        assert_equal [r config get moduleconfigs.numeric] "moduleconfigs.numeric -1"
+        r module unload moduleconfigs
+    }
+
+    test {test loadex functionality} {
+        r module loadex $testmodule CONFIG moduleconfigs.mutable_bool no CONFIG moduleconfigs.immutable_bool yes CONFIG moduleconfigs.memory_numeric 2mb CONFIG moduleconfigs.string tclortickle
+        assert_equal [lindex [lindex [r module list] 0] 1] moduleconfigs
+        assert_equal [r config get moduleconfigs.mutable_bool] "moduleconfigs.mutable_bool no"
+        assert_equal [r config get moduleconfigs.immutable_bool] "moduleconfigs.immutable_bool yes"
+        assert_equal [r config get moduleconfigs.memory_numeric] "moduleconfigs.memory_numeric 2097152"
+        assert_equal [r config get moduleconfigs.string] "moduleconfigs.string tclortickle"
+        # Configs that were not changed should still be their module specified value
+        assert_equal [r config get moduleconfigs.enum] "moduleconfigs.enum one"
+        assert_equal [r config get moduleconfigs.numeric] "moduleconfigs.numeric -1"
+    }
+
+    test {apply function works} {
+        catch {[r config set moduleconfigs.mutable_bool yes]} e
+        assert_match {*Bool configs*} $e
+        assert_equal [r config get moduleconfigs.mutable_bool] "moduleconfigs.mutable_bool no"
+        catch {[r config set moduleconfigs.memory_numeric 1000 moduleconfigs.numeric 1000]} e
+        assert_match {*cannot equal*} $e
+        assert_equal [r config get moduleconfigs.memory_numeric] "moduleconfigs.memory_numeric 2097152"
+        assert_equal [r config get moduleconfigs.numeric] "moduleconfigs.numeric -1"
+        r module unload moduleconfigs
+    }
+
+    test {test loadex rejects bad configs} {
+        # Bad config 200gb is over the limit
```

Comment:
Let's test string as well, since we know there are some memory issues handling constraints.